### ISSUE TITLE
In ReconnectingClientFactory calling stopTrying will sometimes trigger a retry on clientConnectionFailed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4846]
|<img alt="rlotun's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rlotun reported|
|-|-|
|Trac ID|trac#4846|
|Type|defect|
|Created|2011-02-02 18:25:28Z|
|Branch|https://github.com/twisted/twisted/tree/rcf-retry-4846|

In our buildbot we noticed the following traceback when running our unittests:

```
  factory.stopTrying()
      File "/usr/lib/python2.6/dist-packages/twisted/internet/protocol.py", line 394, in stopTrying
    self.connector.stopConnecting()
      File "/usr/lib/python2.6/dist-packages/twisted/internet/base.py", line 1031, in stopConnecting
    self.transport.failIfNotConnected(error.UserError())
      File "/usr/lib/python2.6/dist-packages/twisted/internet/tcp.py", line 582, in failIfNotConnected
    self.connector.connectionFailed(failure.Failure(err))
      File "/usr/lib/python2.6/dist-packages/twisted/internet/base.py", line 1051, in connectionFailed
    self.factory.clientConnectionFailed(self, reason)
      File "/home/ubuntu/buildbot/slave-ec2-XX-XXX-XX-XX.compute-1.amazonaws.com/rlotun-api-tdweb-removal/build/tdapi/redis.py", line 32, in clientConnectionFailed
    RedisClientFactory.clientConnectionFailed(self, connector, reason)
      File "/usr/lib/python2.6/dist-packages/twisted/internet/protocol.py", line 334, in clientConnectionFailed
    self.retry()
      File "/usr/lib/python2.6/dist-packages/twisted/internet/protocol.py", line 379, in retry
    self._callID = self.clock.callLater(self.delay, reconnector)
```

As part of our `tearDown` we were calling `stopTrying` on a `ReconnectingClientFactory`. After examining the code below it seems that if a connector is active `stopConnecting` will be called, which would trigger a `clientConnectionFailed` which would then trigger a retry **while continueTrying is still set to 1**. A simple fix of moving the `self.continueTrying = 0` line up before calling `stopConnecting` seems to do the trick.

_Offending code block_:
```

    def stopTrying(self):
        """
        Put a stop to any attempt to reconnect in progress.
        """
        # ??? Is this function really stopFactory?
        if self._callID:
            self._callID.cancel()
            self._callID = None
        if self.connector:
            # Hopefully this doesn't just make clientConnectionFailed
            # retry again.
            try:
                self.connector.stopConnecting()
            except error.NotConnectingError:
                pass
        self.continueTrying = 0
```

Attached is a patch which includes a test to reproduce the issue.

Attachments:

* [rcf_stoptrying.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/74f/74f5bc0f81d9b77b3f995b71cac554c5cca5f80f/21559ecf935c60c3c48688f9c3df65c68570f479.diff) (2403 bytes) - added by rlotun on 2011-02-02 18:26:02Z - fix and unit test

<details><summary>Searchable metadata</summary>

```
trac-id__4846 4846
type__defect defect
reporter__rlotun rlotun
priority__normal normal
milestone__ 
branch__branches_rcf_retry_4846 branches/rcf-retry-4846
branch_author__rlotun__exarkun rlotun, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1296671128000000 1296671128000000
changetime__1296769431000000 1296769431000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Thanks!  This one was tricky :/  The fix looks right, but here are some comments mainly about the tests:

  1. There should be a [ReviewProcess#Newsfiles news fragment].
  1. Copyright date on modified files should be updated to include 2011.
  1. Methods should be separated by each other by two blank lines
  1. `attempted_retry` should be `attemptedRetry` to follow the variable naming convention
  1. `assertEquals(x, 0)` is better than `assertTrue(x == 0)` since it will display the value of `x` if the assertion fails
  1. The test could avoid overriding any `ReconnectingClientFactory` methods if it overrode `connect` on the fake connector instead.  This would also ensure that the `connect` method isn't called by any other part of `ReconnectingClientFactory`.

These are mostly simple, so I'll implement them.  The last one is pretty easy too (now that I've actually managed to understand what's going on with RFC ;), so I can implement that too, but involved enough that I'm going to ask someone else to review it too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [30602]) Branching to 'rcf-retry-4846'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [30603]) Apply rcf_stoptrying.diff

refs [#4846](https://github.com/twisted/twisted/issues/4846)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [30604]) Adjust test, removing ReconnectingClientFactory subclass and fixing some other style issues

refs [#4846](https://github.com/twisted/twisted/issues/4846)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [30607]) Add some docstrings to make the test a little easier to understand; also delete a Hopeful but naive comment which is no longer relevant.

refs [#4846](https://github.com/twisted/twisted/issues/4846)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I did those things I said.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/rcf-retry-4846)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @exarkun|
|-|-|

The branch looks good to me.  Please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [30609]) Merge rcf-retry-4846

Author: rlotun, exarkun
Reviewer: exarkun, glyph
Fixes: [#4846](https://github.com/twisted/twisted/issues/4846)

Fix a bug in ReconnectingClientFactory which would cause it to sometimes re-try
a connection attempt after its `stopTrying` method was called.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
